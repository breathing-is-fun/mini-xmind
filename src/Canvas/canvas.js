import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classNames from 'classNames';

import Draggable from 'react-draggable';
import { SteppedLineTo } from '../Line';
import { getPlacement, preventDefault, generateKey } from '../utils/LineUtil';
import Block from '../tools/Block';

export default class Canvas extends Component {
  static propTypes = {
    style: PropTypes.object,
    className: PropTypes.string,
  };

  static defaultProps = {
    style: {},
    className: '',
  };

  constructor(props) {
    super(props);

    this.state = {
      blockProps: {},
      lines: {},
    };

    // for Line, two point create a line
    this.checkBlockClickList = {};
    this.blockDOM = {};
  }

  saveBlock = (ref, blockKey) => {
    this.blockDOM[blockKey] = ReactDOM.findDOMNode(ref);
  };

  handleStop = ({ x, y }, blockKey) => {
    const { blockProps } = this.state;

    blockProps[blockKey] = Object.assign({}, blockProps[blockKey], { x, y });

    this.setState({ blockProps });
  };

  onDrop = e => {
    let dragItem = e.dataTransfer.getData('dragItem');
    dragItem = dragItem ? JSON.parse(dragItem) : {};
    const { value, style } = dragItem;
    let { blockProps } = this.state;
    const { clientX, clientY } = e;
    const blockKey = generateKey('block');
    const x = clientX - style.width / 2;
    const y = clientY - style.height / 2;

    switch (value) {
      case 'block':
        blockProps[blockKey] = { x, y, style };
        break;

      case 'line':
        break;

      default:
        break;
    }
    this.setState({ blockProps });
  };

  // to repaint Line instantly
  handleBlockDrag = () => {
    this.setState({});
  };

  // when Block clicked twice, generate a Line
  // and clear checkBlockClickList
  handleBlockClick = blockKey => {
    let { checkBlockClickList, blockDOM } = this;
    const { blockProps } = this.state;

    checkBlockClickList[blockKey] = { current: blockDOM[blockKey] };

    // to know which block is starting point
    if (!('time' in checkBlockClickList[blockKey])) {
      checkBlockClickList[blockKey].time = new Date().getTime();
    }

    if (Object.keys(checkBlockClickList).length == 2) {
      let fromNode, toNode, fromKey, toKey;
      let { lines } = this.state;
      const keys = Object.keys(checkBlockClickList);
      const lineKey = generateKey('line');

      if (checkBlockClickList[keys[0]] > checkBlockClickList[keys[1]]) {
        fromKey = keys[1];
        toKey = keys[0];
      } else {
        fromKey = keys[0];
        toKey = keys[1];
      }

      fromNode = checkBlockClickList[fromKey].current;
      toNode = checkBlockClickList[toKey].current;

      const { fromAnchor, toAnchor } = getPlacement(
        blockProps[fromKey],
        blockProps[toKey],
      );

      lines[lineKey] = {
        from: fromNode,
        to: toNode,
        fromAnchor,
        toAnchor,
        key: lineKey,
      };

      this.setState({ lines }, () => {
        this.checkBlockClickList = {};
      });
    }
  };

  generateBlocks = blockProps => {
    return Object.keys(blockProps).map(blockKey => {
      const { x, y, style } = blockProps[blockKey];

      return (
        <Draggable
          onStop={(e, item) => this.handleStop(item, blockKey)}
          key={blockKey}
          position={{ x, y }}
          onDrag={this.handleBlockDrag}
        >
          <Block
            style={style}
            onClick={e => this.handleBlockClick(blockKey)}
            ref={ref => this.saveBlock(ref, blockKey)}
          />
        </Draggable>
      );
    });
  };

  generateLines = lines => {
    return Object.keys(lines).map(lineKey => {
      const { from, to, key, fromAnchor, toAnchor } = lines[lineKey];

      return (
        <SteppedLineTo
          from={from}
          to={to}
          fromAnchor={fromAnchor}
          toAnchor={toAnchor}
          key={key}
        />
      );
    });
  };

  render = () => {
    const { className, ...rest } = this.props;
    const { blockProps, lines } = this.state;

    return (
      <div
        className={classNames('Canvas', className)}
        onDragOver={preventDefault}
        onDrop={this.onDrop}
        {...rest}
      >
        {this.generateBlocks(blockProps)}
        {this.generateLines(lines)}
      </div>
    );
  };
}
