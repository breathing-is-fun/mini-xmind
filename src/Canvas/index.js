import Canvas from './core';

export default Canvas;
