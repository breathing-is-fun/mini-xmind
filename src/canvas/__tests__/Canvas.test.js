import React from 'react';
import { mount } from 'enzyme';
import Canvas from '..';
import 'nino-cli/scripts/setup';

const style = {
  width: 100,
  height: 80,
  background: '#F96',
  borderRadius: '10px',
  border: '1px solid #aaa',
};
const blockProps = {
  test1: {
    x: 90,
    y: 50,
    style,
  },
  test2: {
    x: 20,
    y: 150,
    style,
  },
  test3: {
    x: 100,
    y: 100,
    style,
  },
};

const createWrapper = () =>
  mount(<Canvas style={{ width: '100%', height: '100%' }} />);

describe('Canvas', () => {
  it('Canvas renders correctly', () => {
    const wrapper = createWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('Block renders correctly', () => {
    const wrapper = createWrapper();
    wrapper.setState({ blockProps });

    expect(wrapper.find('.Block').length).toBe(3);
    expect(wrapper).toMatchSnapshot();
  });

  it('when Block is clicked, Line renders correctly', () => {
    const wrapper = createWrapper();
    wrapper.setState({ blockProps });

    const blocks = wrapper.find('.Block');
    blocks.at(0).simulate('click');
    blocks.at(1).simulate('click');

    blocks.at(1).simulate('click');
    blocks.at(2).simulate('click');

    blocks.at(0).simulate('click');
    blocks.at(2).simulate('click');
    expect(wrapper.find('.stepped-line-to').length).toBe(3);
  });

  it('should not render redundant Line', () => {
    const wrapper = createWrapper();
    wrapper.setState({ blockProps });

    const blocks = wrapper.find('.Block');
    blocks.at(0).simulate('click');
    blocks.at(1).simulate('click');
    expect(wrapper.find('.stepped-line-to').length).toBe(1);

    blocks.at(0).simulate('click');
    blocks.at(1).simulate('click');
    expect(wrapper.find('.stepped-line-to').length).toBe(1);
  });
});
