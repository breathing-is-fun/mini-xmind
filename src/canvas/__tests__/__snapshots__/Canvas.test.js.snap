// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas Block renders correctly 1`] = `
<Canvas
  className=""
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <div
    className="Canvas"
    onDragOver={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <Block
      className=""
      data={
        Object {
          "test1": Object {
            "style": Object {
              "background": "#F96",
              "border": "1px solid #aaa",
              "borderRadius": "10px",
              "height": 80,
              "width": 100,
            },
            "x": 90,
            "y": 50,
          },
          "test2": Object {
            "style": Object {
              "background": "#F96",
              "border": "1px solid #aaa",
              "borderRadius": "10px",
              "height": 80,
              "width": 100,
            },
            "x": 20,
            "y": 150,
          },
          "test3": Object {
            "style": Object {
              "background": "#F96",
              "border": "1px solid #aaa",
              "borderRadius": "10px",
              "height": 80,
              "width": 100,
            },
            "x": 100,
            "y": 100,
          },
        }
      }
      lineData={Object {}}
      onChange={[Function]}
    >
      <Draggable
        allowAnyClick={false}
        axis="both"
        bounds={false}
        cancel={null}
        defaultClassName="react-draggable"
        defaultClassNameDragged="react-draggable-dragged"
        defaultClassNameDragging="react-draggable-dragging"
        defaultPosition={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        disabled={false}
        enableUserSelectHack={true}
        grid={null}
        handle={null}
        key="test1"
        offsetParent={null}
        onDrag={[Function]}
        onMouseDown={[Function]}
        onStart={[Function]}
        onStop={[Function]}
        position={
          Object {
            "x": 90,
            "y": 50,
          }
        }
        scale={1}
        transform={null}
      >
        <DraggableCore
          allowAnyClick={false}
          axis="both"
          bounds={false}
          cancel={null}
          defaultClassName="react-draggable"
          defaultClassNameDragged="react-draggable-dragged"
          defaultClassNameDragging="react-draggable-dragging"
          defaultPosition={
            Object {
              "x": 0,
              "y": 0,
            }
          }
          disabled={false}
          enableUserSelectHack={true}
          grid={null}
          handle={null}
          offsetParent={null}
          onDrag={[Function]}
          onMouseDown={[Function]}
          onStart={[Function]}
          onStop={[Function]}
          position={
            Object {
              "x": 90,
              "y": 50,
            }
          }
          scale={1}
          transform={null}
        >
          <div
            className="BlockGroup react-draggable"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "background": "#F96",
                "border": "1px solid #aaa",
                "borderRadius": "10px",
                "height": 80,
                "touchAction": "none",
                "transform": "translate(90px,50px)",
                "width": 100,
              }
            }
            transform={null}
          />
        </DraggableCore>
      </Draggable>
      <Draggable
        allowAnyClick={false}
        axis="both"
        bounds={false}
        cancel={null}
        defaultClassName="react-draggable"
        defaultClassNameDragged="react-draggable-dragged"
        defaultClassNameDragging="react-draggable-dragging"
        defaultPosition={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        disabled={false}
        enableUserSelectHack={true}
        grid={null}
        handle={null}
        key="test2"
        offsetParent={null}
        onDrag={[Function]}
        onMouseDown={[Function]}
        onStart={[Function]}
        onStop={[Function]}
        position={
          Object {
            "x": 20,
            "y": 150,
          }
        }
        scale={1}
        transform={null}
      >
        <DraggableCore
          allowAnyClick={false}
          axis="both"
          bounds={false}
          cancel={null}
          defaultClassName="react-draggable"
          defaultClassNameDragged="react-draggable-dragged"
          defaultClassNameDragging="react-draggable-dragging"
          defaultPosition={
            Object {
              "x": 0,
              "y": 0,
            }
          }
          disabled={false}
          enableUserSelectHack={true}
          grid={null}
          handle={null}
          offsetParent={null}
          onDrag={[Function]}
          onMouseDown={[Function]}
          onStart={[Function]}
          onStop={[Function]}
          position={
            Object {
              "x": 20,
              "y": 150,
            }
          }
          scale={1}
          transform={null}
        >
          <div
            className="BlockGroup react-draggable"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "background": "#F96",
                "border": "1px solid #aaa",
                "borderRadius": "10px",
                "height": 80,
                "touchAction": "none",
                "transform": "translate(20px,150px)",
                "width": 100,
              }
            }
            transform={null}
          />
        </DraggableCore>
      </Draggable>
      <Draggable
        allowAnyClick={false}
        axis="both"
        bounds={false}
        cancel={null}
        defaultClassName="react-draggable"
        defaultClassNameDragged="react-draggable-dragged"
        defaultClassNameDragging="react-draggable-dragging"
        defaultPosition={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        disabled={false}
        enableUserSelectHack={true}
        grid={null}
        handle={null}
        key="test3"
        offsetParent={null}
        onDrag={[Function]}
        onMouseDown={[Function]}
        onStart={[Function]}
        onStop={[Function]}
        position={
          Object {
            "x": 100,
            "y": 100,
          }
        }
        scale={1}
        transform={null}
      >
        <DraggableCore
          allowAnyClick={false}
          axis="both"
          bounds={false}
          cancel={null}
          defaultClassName="react-draggable"
          defaultClassNameDragged="react-draggable-dragged"
          defaultClassNameDragging="react-draggable-dragging"
          defaultPosition={
            Object {
              "x": 0,
              "y": 0,
            }
          }
          disabled={false}
          enableUserSelectHack={true}
          grid={null}
          handle={null}
          offsetParent={null}
          onDrag={[Function]}
          onMouseDown={[Function]}
          onStart={[Function]}
          onStop={[Function]}
          position={
            Object {
              "x": 100,
              "y": 100,
            }
          }
          scale={1}
          transform={null}
        >
          <div
            className="BlockGroup react-draggable"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "background": "#F96",
                "border": "1px solid #aaa",
                "borderRadius": "10px",
                "height": 80,
                "touchAction": "none",
                "transform": "translate(100px,100px)",
                "width": 100,
              }
            }
            transform={null}
          />
        </DraggableCore>
      </Draggable>
    </Block>
    <LineGroup
      data={Object {}}
    >
      <div
        className="LineGroup"
      />
    </LineGroup>
    <TagGroup
      data={Object {}}
      onChange={[Function]}
    />
  </div>
</Canvas>
`;

exports[`Canvas Canvas renders correctly 1`] = `
<Canvas
  className=""
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <div
    className="Canvas"
    onDragOver={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <Block
      className=""
      data={Object {}}
      lineData={Object {}}
      onChange={[Function]}
    />
    <LineGroup
      data={Object {}}
    >
      <div
        className="LineGroup"
      />
    </LineGroup>
    <TagGroup
      data={Object {}}
      onChange={[Function]}
    />
  </div>
</Canvas>
`;
